File: internal/collector/model/alert.go
// Copyright 2024 ff, Scalytics, Inc. - https://www.scalytics.io
// SPDX-License-Identifier: Apache-2.0

package model

type Alert struct {
	AlertID            string            `json:"alert_id"`
	SourceID           string            `json:"source_id"`
	Source             SourceMetadata    `json:"source"`
	Title              string            `json:"title"`
	CanonicalURL       string            `json:"canonical_url"`
	FirstSeen          string            `json:"first_seen"`
	LastSeen           string            `json:"last_seen"`
	Status             string            `json:"status"`
	Category           string            `json:"category"`
	Severity           string            `json:"severity"`
	SignalLane         SignalLane        `json:"signal_lane,omitempty"`
	RegionTag          string            `json:"region_tag"`
	Lat                float64           `json:"lat"`
	Lng                float64           `json:"lng"`
	EventCountry       string            `json:"event_country,omitempty"`
	EventCountryCode   string            `json:"event_country_code,omitempty"`
	EventGeoSource     string            `json:"event_geo_source,omitempty"`
	EventGeoConfidence float64           `json:"event_geo_confidence,omitempty"`
	FreshnessHours     int               `json:"freshness_hours"`
	Reporting          ReportingMetadata `json:"reporting,omitempty"`
	Triage             *Triage           `json:"triage,omitempty"`
}

type SignalLane string

const (
	SignalLaneAlarm SignalLane = "alarm"
	SignalLaneIntel SignalLane = "intel"
	SignalLaneInfo  SignalLane = "info"
)

type Triage struct {
	RelevanceScore  float64         `json:"relevance_score"`
	Threshold       float64         `json:"threshold,omitempty"`
	Confidence      string          `json:"confidence,omitempty"`
	Disposition     string          `json:"disposition,omitempty"`
	PublicationType string          `json:"publication_type,omitempty"`
	WeakSignals     []string        `json:"weak_signals,omitempty"`
	Metadata        *TriageMetadata `json:"metadata,omitempty"`
	Reasoning       string          `json:"reasoning,omitempty"`
}

type TriageMetadata struct {
	Author                 string   `json:"author,omitempty"`
	Tags                   []string `json:"tags,omitempty"`
	NoiseDecision          string   `json:"noise_decision,omitempty"`
	NoisePolicyVersion     string   `json:"noise_policy_version,omitempty"`
	NoisePolicyVariant     string   `json:"noise_policy_variant,omitempty"`
	NoiseBlockScore        float64  `json:"noise_block_score,omitempty"`
	NoiseScore             float64  `json:"noise_score,omitempty"`
	NoiseActionability     float64  `json:"noise_actionability_score,omitempty"`
	NoiseReasons           []string `json:"noise_reasons,omitempty"`
	NoiseDecisionTimestamp string   `json:"noise_decision_ts,omitempty"`
}

type DuplicateSample struct {
	Title string `json:"title"`
	Count int    `json:"count"`
}

type DuplicateAudit struct {
	SuppressedVariantDuplicates int               `json:"suppressed_variant_duplicates"`
	RepeatedTitleGroupsInActive int               `json:"repeated_title_groups_in_active"`
	RepeatedTitleSamples        []DuplicateSample `json:"repeated_title_samples"`
}

type SourceHealthEntry struct {
	SourceID         string `json:"source_id"`
	AuthorityName    string `json:"authority_name"`
	Type             string `json:"type"`
	Status           string `json:"status"`
	FetchedCount     int    `json:"fetched_count"`
	FeedURL          string `json:"feed_url"`
	Error            string `json:"error,omitempty"`
	ErrorClass       string `json:"error_class,omitempty"`
	NeedsReplacement bool   `json:"needs_replacement,omitempty"`
	DiscoveryAction  string `json:"discovery_action,omitempty"`
	StartedAt        string `json:"started_at"`
	FinishedAt       string `json:"finished_at"`
	ActiveCount      int    `json:"active_count,omitempty"`
	FilteredCount    int    `json:"filtered_count,omitempty"`
	// Cache metadata — not serialised to JSON, used internally to
	// populate source_watermarks so the next cycle can do conditional GET.
	RespETag         string `json:"-"`
	RespLastModified string `json:"-"`
}

type SourceReplacementCandidate struct {
	SourceID        string `json:"source_id"`
	AuthorityName   string `json:"authority_name"`
	Type            string `json:"type"`
	FeedURL         string `json:"feed_url"`
	BaseURL         string `json:"base_url,omitempty"`
	Country         string `json:"country,omitempty"`
	CountryCode     string `json:"country_code,omitempty"`
	Region          string `json:"region,omitempty"`
	AuthorityType   string `json:"authority_type,omitempty"`
	Category        string `json:"category,omitempty"`
	Error           string `json:"error,omitempty"`
	ErrorClass      string `json:"error_class,omitempty"`
	DiscoveryAction string `json:"discovery_action,omitempty"`
	LastAttemptAt   string `json:"last_attempt_at,omitempty"`
}

type SourceHealthDocument struct {
	GeneratedAt             string                       `json:"generated_at"`
	CriticalSourcePrefixes  []string                     `json:"critical_source_prefixes"`
	FailOnCriticalSourceGap bool                         `json:"fail_on_critical_source_gap"`
	TotalSources            int                          `json:"total_sources"`
	SourcesOK               int                          `json:"sources_ok"`
	SourcesError            int                          `json:"sources_error"`
	DuplicateAudit          DuplicateAudit               `json:"duplicate_audit"`
	ReplacementQueue        []SourceReplacementCandidate `json:"replacement_queue"`
	Sources                 []SourceHealthEntry          `json:"sources"`
}

type SourceReplacementDocument struct {
	GeneratedAt string                       `json:"generated_at"`
	Sources     []SourceReplacementCandidate `json:"sources"`
}

type SourceCandidate struct {
	URL           string `json:"url"`
	AuthorityName string `json:"authority_name,omitempty"`
	AuthorityType string `json:"authority_type,omitempty"`
	Category      string `json:"category,omitempty"`
	Country       string `json:"country,omitempty"`
	CountryCode   string `json:"country_code,omitempty"`
	Region        string `json:"region,omitempty"`
	BaseURL       string `json:"base_url,omitempty"`
	Notes         string `json:"notes,omitempty"`
}

type SourceCandidateDocument struct {
	GeneratedAt string            `json:"generated_at,omitempty"`
	Sources     []SourceCandidate `json:"sources"`
}

File: src/types/alert.ts
/*
 * EUOSINT
 * Portions derived from novatechflow/osint-siem and cyberdude88/osint-siem.
 * See NOTICE for provenance and LICENSE for repository-local terms.
 */

export type Severity = "critical" | "high" | "medium" | "low" | "info";
export type AlertStatus = "active" | "updated" | "removed" | "filtered";
export type SignalLane = "alarm" | "intel" | "info";
export type AlertCategory =
  | "informational"
  | "missing_person"
  | "wanted_suspect"
  | "public_appeal"
  | "cyber_advisory"
  | "education_digital_capacity"
  | "humanitarian_tasking"
  | "conflict_monitoring"
  | "humanitarian_security"
  | "terrorism_tip"
  | "fraud_alert"
  | "public_safety"
  | "private_sector"
  | "travel_warning"
  | "health_emergency"
  | "intelligence_report"
  | "emergency_management"
  | "environmental_disaster"
  | "disease_outbreak"
  | "maritime_security"
  | "logistics_incident"
  | "legislative";
export type AuthorityType =
  | "police"
  | "national_security"
  | "intelligence"
  | "regulatory"
  | "public_safety_program"
  | "cert"
  | "private_sector"
  | "osint";

export interface AuthoritySource {
  source_id: string;
  authority_name: string;
  country: string;
  country_code: string;
  region: string;
  authority_type: AuthorityType;
  base_url: string;
}

export interface Alert {
  alert_id: string;
  source_id: string;
  source: AuthoritySource;
  title: string;
  canonical_url: string;
  first_seen: string;
  last_seen: string;
  status: AlertStatus;
  category: AlertCategory;
  severity: Severity;
  signal_lane?: SignalLane;
  region_tag: string;
  lat: number;
  lng: number;
  event_country?: string;
  event_country_code?: string;
  event_geo_source?: string;
  event_geo_confidence?: number;
  freshness_hours: number;
  reporting?: ReportingInfo;
  triage?: AlertTriage;
}

export interface ReportingInfo {
  label: string;
  url?: string;
  phone?: string;
  email?: string;
  notes?: string;
}

export interface AlertTriage {
  relevance_score: number;
  threshold: number;
  confidence: "high" | "medium" | "low";
  disposition: "retained" | "filtered_review";
  publication_type?: string;
  weak_signals?: string[];
  metadata?: {
    author?: string;
    tags?: string[];
  };
}
